pkg/service/init.go

```diff
@@ -38,39 +36,35 @@ func NewBootstrap(router *mux.Router) *Bootstrap {
 }
 
 func (b *Bootstrap) BootstrapHandler(ctx context.Context, wg *sync.WaitGroup, startupTimer startup.Timer, dic *di.Container) (success bool) {
-	// TODO: remove these after refactor are done.
-	common.CurrentConfig = container.ConfigurationFrom(dic.Get)
-	common.LoggingClient = bootstrapContainer.LoggingClientFrom(dic.Get)
-	common.RegistryClient = bootstrapContainer.RegistryFrom(dic.Get)
-
 	ds.UpdateFromContainer(b.router, dic)
-	ds.controller.InitRestRoutes(dic)
-
-	lc := ds.LoggingClient
-	configuration := ds.config
 	autoevent.NewManager(ctx, wg)
 
+	err := ds.selfRegister()
+	if err != nil {
+		ds.LoggingClient.Error(fmt.Sprintf("Couldn't register to metadata service: %v\n", err))
+		return false
+	}
+
+	// initialize devices, deviceResources, provisionWatchers & profiles cache
+	cache.InitCache(
+		ds.ServiceName,
+		ds.LoggingClient,
+		container.CoredataValueDescriptorClientFrom(dic.Get),
+		container.MetadataDeviceClientFrom(dic.Get),
+		container.MetadataProvisionWatcherClientFrom(dic.Get))
+
```

Comment:
Why did `selfRegister` and `InitCache` get moved earlier in the startup sequence?

Reply:
I tried to init `asyncCh` and `deviceCh` before where they're actually used (i.e. pass into ProtocolDriver initialization) so the `selfRegister` and `InitCache` got moved earlier as a result. If this is unnecessary I will change it back.

Reply:
OK, thanks for the explanation. After going back over the code, I think this will be OK. I assume you've tested some of the existing device services with this change already? If not, please make sure you do so before merging.